fix buggy logic in type resolution of type suffixes. #34

SSoulaimane opened the pull request:
issue #34

BBasile commented on src/dsymbol/tests.d:
```diff
@@ -66,3 +69,77 @@ unittest
+
+    auto tokens = lex(q{int a = 9;});
+    foreach(i, t;
+        AliasSeq!(tok!"int", tok!"identifier", tok!"=", tok!"intLiteral", tok!";"))
```
You can use an array literal here, there is no need for the values to be available at compile time

BBasile replied:
Dismiss this, since this comment i suggest to change the tests in another PR

SSoulaimane replied:
sure

BBasile commented on src/dsymbol/tests.d:
```diff
@@ -66,3 +69,77 @@ unittest
+{
+    import std.uuid : randomUUID;
+    import std.string : translate;
+    return "dsymbol_" ~ randomUUID().toString().translate(['-': '_']) ~ ".d";
```
you don't need to translate.

BBasile replied:
Dismiss this, since this comment i suggest to change the tests in another PR

BBasile commented on src/dsymbol/tests.d:
```diff
@@ -15,7 +18,7 @@ import std.stdio : writeln, stdout;
+version (unittest):
 void expectSymbolsAndTypes(const string source, const string[][] results)
 {
     static immutable rName = "dsymbol-test-154251542-56564105-78944416-98523170-02452336.d";
```
why not replacing this with a `randomDFilename()` ?

BBasile replied:
Dismiss this, since this comment i suggest to change the tests in another PR

BBasile commented on src/dsymbol/conversion/second.d:
```diff
@@ -298,6 +301,53 @@ body
 	}
 }
 
+unittest
```
Why don't you use the existing testing system ? It's more compact and does the same.

SSoulaimane replied:
True. I like the declarative approach, it is more compact.
but the alternative is more flexible, more things are testable for potential bugs, for instance: symbol.kind, symbol.qualifier , symbol.isPointer ...etc.
Maybe we should see if we can make it the current system more flexible.

BBasile commented:
#36 passes, let's find why. The only difference is that i use the existing tester routine.

BBasile commented:
#37 created a conflict and forces you to rebase and remove your tests. I've even added the tests for POINTER_SYMBOL_NAME, which just have to be un-commented.

BBasile commented:
There's a serious problem with this PR actually, it breaks completion of pointers in DCD. For now DCD follows D implicit dereference. You can see the effect with   ```int* a.|```. Before the change we got completion for an int, after we got nothing.

The right behavior would be to ignore the first POINTER_SYMBOL_NAME since for example even the compiler allows this  `int* a; writeln(a.max)`. If you add an indirection it doesn't work anymore. Currently D symbol always gives the completions, whatever is the indirections count.

SSoulaimane commented:
> #36 passes, let's find why. The only difference is that i use the existing tester routine.
that is one weird bug that apparently only affects dmd >= 2.075.
i got the solution/workaround though.

BBasile commented:
You didn't address the issue that is the that first indirection should not cause DCD not to give the completions. Try your modified file in DCD and you'll see that 

```d
struct Foo{int a;}
Foo* foo;
foo. // dot completion here
```

does not give results anymore.

SSoulaimane commented:
> You didn't address the issue that is the that first indirection should not cause DCD not to give the completions.

I'm definitely going to fix that in dcd. its should be trivial. I've been doing a lot of changes in my local dcd fork, among these some refactoring of symbol swapping with symbol.type which is the real cause of this bug, also added swapping eponymous templates with their content.

the reason I need pointers to be distinguished is because I've been progressing with type checking in dcd for ufcs but the code relies on many changes to dsymbol which I would like to push first.

I there's a few bugs I've found which some I've fixed locally and some not, I'll report them one by one and push the fixes with some tests.

BBasile commented:
Looks like it's in a dead end, closing.
